### ISSUE TITLE
Issue with setting credentials: server error 500

### ISSUE THREAD

**Issue by stefanproell** [jupyter/docker-stacks#639]
Hello,
I am trying to launch a Jupyter container, but after providing the token and the password, there occurs a server error. The file cannot be found

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/jovyan/.jupyter/jupyter_notebook_config.json'
```

Affected image

```
jupyter/datascience-notebook:4ebeb1f2d154
jupyter/minimal-notebook:4ebeb1f2d154
```
## These are the steps:

1. Launch a container with `docker run -p 8888:8888 --name data_science --rm jupyter/minimal-notebook:4ebeb1f2d154`
2. Wait until the token link is shown in the log
3. Open the page in the browser
4. I am directed to the Jupyter directory tree and click on a folder. I am taken to the login page
5. Enter token and a new password
6. Error 500 page is shown

## Error message

The log shows the following error:

```
docker run -p 8888:8888 --name data_science --rm jupyter/minimal-notebook:4ebeb1f2d154
/usr/local/bin/start-notebook.sh: ignoring /usr/local/bin/start-notebook.d/*

Container must be run with group root to update passwd file
Executing the command: jupyter notebook
[I 09:43:46.845 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 09:43:47.052 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[I 09:43:47.085 NotebookApp] JupyterLab beta preview extension loaded from /opt/conda/lib/python3.6/site-packages/jupyterlab
[I 09:43:47.085 NotebookApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 09:43:47.092 NotebookApp] Serving notebooks from local directory: /home/jovyan
[I 09:43:47.092 NotebookApp] 0 active kernels
[I 09:43:47.092 NotebookApp] The Jupyter Notebook is running at:
[I 09:43:47.092 NotebookApp] http://[all ip addresses on your system]:8888/?token=cc7c860bdf3de857f55f012dfab7e86ec3a435d769a7a24b
[I 09:43:47.092 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 09:43:47.092 NotebookApp] 
    
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://localhost:8888/?token=cc7c860bdf3de857f55f012dfab7e86ec3a435d769a7a24b
[I 09:43:50.866 NotebookApp] 302 GET /?token=cc7c860bdf3de857f55f012dfab7e86ec3a435d769a7a24b (172.17.0.1) 0.74ms
[W 09:43:51.387 NotebookApp] Forbidden
[W 09:43:51.388 NotebookApp] 403 GET /api/config/tree?_=1526982231130 (172.17.0.1) 1.32ms referer=http://localhost:8888/tree
[W 09:43:51.389 NotebookApp] Forbidden
[W 09:43:51.390 NotebookApp] 403 GET /api/config/common?_=1526982231131 (172.17.0.1) 0.92ms referer=http://localhost:8888/tree
[W 09:43:51.416 NotebookApp] Forbidden
[W 09:43:51.416 NotebookApp] 403 GET /api/terminals?_=1526982231132 (172.17.0.1) 1.15ms referer=http://localhost:8888/tree
[W 09:43:51.419 NotebookApp] Forbidden
[W 09:43:51.419 NotebookApp] 403 GET /api/kernelspecs (172.17.0.1) 1.02ms referer=http://localhost:8888/tree
[W 09:43:51.420 NotebookApp] Forbidden
[W 09:43:51.421 NotebookApp] 403 GET /api/sessions?_=1526982231133 (172.17.0.1) 0.93ms referer=http://localhost:8888/tree
[W 09:43:51.422 NotebookApp] Forbidden
[W 09:43:51.422 NotebookApp] 403 GET /api/terminals?_=1526982231134 (172.17.0.1) 0.95ms referer=http://localhost:8888/tree
[I 09:43:52.816 NotebookApp] 302 GET /tree (172.17.0.1) 0.64ms
[E 09:44:12.761 NotebookApp] Uncaught exception POST /login?next=%2Ftree (172.17.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='POST', uri='/login?next=%2Ftree', version='HTTP/1.1', remote_ip='172.17.0.1')
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/tornado/web.py", line 1541, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/login.py", line 88, in post
        set_password(new_password, config_file=config_file)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/security.py", line 148, in set_password
        config.NotebookApp.password = hashed_password
      File "/opt/conda/lib/python3.6/contextlib.py", line 88, in __exit__
        next(self.gen)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/security.py", line 131, in persist_config
        with io.open(config_file, 'w', encoding='utf8') as f:
    FileNotFoundError: [Errno 2] No such file or directory: '/home/jovyan/.jupyter/jupyter_notebook_config.json'
[E 09:44:12.778 NotebookApp] {
      "Host": "localhost:8888",
      "Connection": "keep-alive",
      "Content-Length": "144",
      "Cache-Control": "max-age=0",
      "Origin": "http://localhost:8888",
      "Upgrade-Insecure-Requests": "1",
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
      "Referer": "http://localhost:8888/login?next=%2Ftree",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "en-US,en;q=0.9,de;q=0.8",
      "Cookie": "JSESSIONID=EDA359FFD9E863E42A3F4E9547A740E2; _xsrf=2|6612760b|79087624cf3ce0fabd60e41704754d33|1526978520"
    }
[E 09:44:12.778 NotebookApp] 500 POST /login?next=%2Ftree (172.17.0.1) 18.11ms referer=http://localhost:8888/login?next=%2Ftree
```
This seems to be a bug.

**Comment by parente** [comment]
Thanks for the report @stefanproell . I'm not able to reproduce the problem you're seeing after pulling the same image and running the same command:

```
❯❯❯ docker pull jupyter/minimal-notebook:4ebeb1f2d154
4ebeb1f2d154: Pulling from jupyter/minimal-notebook
22dc81ace0ea: Already exists
1a8b3c87dba3: Already exists
91390a1c435a: Already exists
07844b14977e: Already exists
b78396653dae: Already exists
f6ed4d15e19b: Already exists
4b867dd0bba7: Already exists
2bc99ec6f992: Already exists
7037e70e2c06: Already exists
33568953b30a: Already exists
037d62f84cd0: Already exists
7440b3093c97: Already exists
1d38a9cc5843: Already exists
06128c9195aa: Already exists
8dcbf63f1bdd: Already exists
ac973d449e53: Already exists
295d709fd9d2: Already exists
e819d3b4d60a: Already exists
35996541f5ca: Pull complete
Digest: sha256:c64b3eb79320b505734e380f1d6bf4c967adebc426c95e11786d05265eea51bc
Status: Downloaded newer image for jupyter/minimal-notebook:4ebeb1f2d154
❯❯❯ docker run -p 8888:8888 --name data_science --rm jupyter/minimal-notebook:4ebeb1f2d154

/usr/local/bin/start-notebook.sh: ignoring /usr/local/bin/start-notebook.d/*

Container must be run with group root to update passwd file
Executing the command: jupyter notebook
[I 03:03:36.986 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 03:03:38.995 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[I 03:03:39.165 NotebookApp] JupyterLab beta preview extension loaded from /opt/conda/lib/python3.6/site-packages/jupyterlab
[I 03:03:39.166 NotebookApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 03:03:39.209 NotebookApp] Serving notebooks from local directory: /home/jovyan
[I 03:03:39.210 NotebookApp] 0 active kernels
[I 03:03:39.211 NotebookApp] The Jupyter Notebook is running at:
[I 03:03:39.211 NotebookApp] http://[all ip addresses on your system]:8888/?token=71877ec0801f8fe39db001defc0437c3c30c759a19d944fd
[I 03:03:39.212 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 03:03:39.214 NotebookApp]

    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://localhost:8888/?token=71877ec0801f8fe39db001defc0437c3c30c759a19d944fd
[I 03:03:44.483 NotebookApp] 302 GET /?token=71877ec0801f8fe39db001defc0437c3c30c759a19d944fd (172.17.0.1) 1.56ms
[I 03:03:53.886 NotebookApp] Creating new notebook in
[I 03:03:53.947 NotebookApp] Writing notebook-signing key to /home/jovyan/.local/share/jupyter/notebook_secret
[W 03:03:54.828 NotebookApp] 404 GET /nbextensions/widgets/notebook/js/extension.js?v=20180523030336 (172.17.0.1) 30.23ms referer=http://localhost:8888/notebooks/Untitled.ipynb?kernel_name=python3
[I 03:03:55.074 NotebookApp] Kernel started: e7303281-8724-4334-b57c-b08d39802286
[I 03:03:55.890 NotebookApp] Adapting to protocol v5.1 for kernel e7303281-8724-4334-b57c-b08d39802286
```

The final log lines there show a kernel launching successfully. I also clicked into the `work/` folder in the file browser without issue.

It's odd that your browser redirected you to enter a password after you visited the token URL. Are you using any other docker arguments to launch the container? Have / had another notebook server running on localhost:8888 that maybe has a cookie or cached in your browser?

**Comment by DimaK415** [comment]
I am having the same issue.  The problem seems to be at line **131 in security.py**

I've attempted to generate the file manually in the dockerfile using both **jupyter notebooks --generate-config** and manually creating the file.  Maybe changing write mode to w+ in security.py would do the trick.
Also worth noting.  I'm not sure if home/jovyan/.jovyan is supposed to be visible (permissions for jovyan), but its not.

**Note**.  This is using base-notebook.

**Comment by stefanproell** [comment]
Sorry for my late reply.

My local Python version is irrelevant, because it is shipped with the Docker container.
The same applies for the rest of the ecosystem.

I just tried again @parente and still have the same results.
The image is up to date

```
	docker pull jupyter/minimal-notebook:4ebeb1f2d154
	4ebeb1f2d154: Pulling from jupyter/minimal-notebook
	Digest: sha256:c64b3eb79320b505734e380f1d6bf4c967adebc426c95e11786d05265eea51bc
	Status: Image is up to date for jupyter/minimal-notebook:4ebeb1f2d154
```
Launch the container:
```
docker run -p 8888:8888 --name data_science --rm jupyter/minimal-notebook:4ebeb1f2d154
/usr/local/bin/start-notebook.sh: ignoring /usr/local/bin/start-notebook.d/*

Container must be run with group root to update passwd file
Executing the command: jupyter notebook
[I 06:53:21.456 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 06:53:21.673 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[I 06:53:21.699 NotebookApp] JupyterLab beta preview extension loaded from /opt/conda/lib/python3.6/site-packages/jupyterlab
[I 06:53:21.699 NotebookApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 06:53:21.705 NotebookApp] Serving notebooks from local directory: /home/jovyan
[I 06:53:21.705 NotebookApp] 0 active kernels
[I 06:53:21.705 NotebookApp] The Jupyter Notebook is running at:
[I 06:53:21.705 NotebookApp] http://[all ip addresses on your system]:8888/?token=8b7ef625824efba3281588ae0cfc031698bb8254bca96c6c
[I 06:53:21.705 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 06:53:21.705 NotebookApp] 
    
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://localhost:8888/?token=8b7ef625824efba3281588ae0cfc031698bb8254bca96c6c
[I 06:53:31.767 NotebookApp] 302 GET /?token=8b7ef625824efba3281588ae0cfc031698bb8254bca96c6c (172.17.0.1) 1.18ms
[W 06:53:32.114 NotebookApp] Forbidden
[W 06:53:32.114 NotebookApp] 403 GET /api/config/tree?_=1529391211941 (172.17.0.1) 0.88ms referer=http://localhost:8888/tree
[W 06:53:32.115 NotebookApp] Forbidden
[W 06:53:32.115 NotebookApp] 403 GET /api/config/common?_=1529391211942 (172.17.0.1) 0.61ms referer=http://localhost:8888/tree
[W 06:53:32.132 NotebookApp] Forbidden
[W 06:53:32.132 NotebookApp] 403 GET /api/terminals?_=1529391211943 (172.17.0.1) 0.71ms referer=http://localhost:8888/tree
[W 06:53:32.134 NotebookApp] Forbidden
[W 06:53:32.134 NotebookApp] 403 GET /api/kernelspecs (172.17.0.1) 0.66ms referer=http://localhost:8888/tree
[W 06:53:32.135 NotebookApp] Forbidden
[W 06:53:32.135 NotebookApp] 403 GET /api/sessions?_=1529391211944 (172.17.0.1) 0.62ms referer=http://localhost:8888/tree
[W 06:53:32.136 NotebookApp] Forbidden
[W 06:53:32.136 NotebookApp] 403 GET /api/terminals?_=1529391211945 (172.17.0.1) 0.64ms referer=http://localhost:8888/tree
```

The error
```
[E 06:53:58.974 NotebookApp] Uncaught exception POST /login?next=%2F (172.17.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='POST', uri='/login?next=%2F', version='HTTP/1.1', remote_ip='172.17.0.1')
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/tornado/web.py", line 1541, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/login.py", line 88, in post
        set_password(new_password, config_file=config_file)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/security.py", line 148, in set_password
        config.NotebookApp.password = hashed_password
      File "/opt/conda/lib/python3.6/contextlib.py", line 88, in __exit__
        next(self.gen)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/security.py", line 131, in persist_config
        with io.open(config_file, 'w', encoding='utf8') as f:
    FileNotFoundError: [Errno 2] No such file or directory: '/home/jovyan/.jupyter/jupyter_notebook_config.json'
[E 06:53:58.984 NotebookApp] {
      "Host": "localhost:8888",
      "Connection": "keep-alive",
      "Content-Length": "143",
      "Cache-Control": "max-age=0",
      "Origin": "http://localhost:8888",
      "Upgrade-Insecure-Requests": "1",
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.79 Safari/537.36",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
      "Referer": "http://localhost:8888/login",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "en-US,en;q=0.9,de;q=0.8",
      "Cookie": "SESSION=d203dd40-7d2a-47df-b278-c0e73f03113a; _xsrf=2|eaa1cef9|2aa8ce0ccfa34f1e71b869a82bdbae6e|1529391211"
```
I tried it with the latest image:

```
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://ad3d38562754:8888/?token=14926ae690bc43993ac18fc4b7a55b193f6f0ce732e8907e&token=14926ae690bc43993ac18fc4b7a55b193f6f0ce732e8907e
```
This image now uses the container name instead of the IP / localhost and repeats the token twice. I can login with the token, upon trying to login with a password again a 500 error appears. The error message is:
```
[E 07:00:28.391 NotebookApp] Uncaught exception POST /login?next=%2F (172.17.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='POST', uri='/login?next=%2F', version='HTTP/1.1', remote_ip='172.17.0.1')
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/tornado/web.py", line 1541, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/login.py", line 88, in post
        set_password(new_password, config_file=config_file)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/security.py", line 148, in set_password
        config.NotebookApp.password = hashed_password
      File "/opt/conda/lib/python3.6/contextlib.py", line 88, in __exit__
        next(self.gen)
      File "/opt/conda/lib/python3.6/site-packages/notebook/auth/security.py", line 131, in persist_config
        with io.open(config_file, 'w', encoding='utf8') as f:
    FileNotFoundError: [Errno 2] No such file or directory: '/home/jovyan/.jupyter/jupyter_notebook_config.json'
[E 07:00:28.398 NotebookApp] {
      "Host": "localhost:8888",
      "Connection": "keep-alive",
      "Content-Length": "143",
      "Cache-Control": "max-age=0",
      "Origin": "http://localhost:8888",
      "Upgrade-Insecure-Requests": "1",
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.79 Safari/537.36",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
      "Referer": "http://localhost:8888/login",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "en-US,en;q=0.9,de;q=0.8",
      "Cookie": "SESSION=d203dd40-7d2a-47df-b278-c0e73f03113a; _xsrf=2|eaa1cef9|2aa8ce0ccfa34f1e71b869a82bdbae6e|1529391211"
    }
```

**Comment by parente** [comment]
@stefanproell can you try running the following:

```
docker run -p 18888:8888 --name data_science --rm jupyter/minimal-notebook:4ebeb1f2d154
```

and visiting http://localhost:18888 instead to see if the problem persists?

**Comment by stefanproell** [comment]
Hello,
yes, the problem also persists with

> http://localhost:18888/login?next=%2F

**Comment by stefanproell** [comment]
Hello,
this also persists with the latest image 3772fffc4aa4.
I really wonder that this does not affect more people.

```
version: '2'
services:
  datascience-notebook:
      image:            jupyter/minimal-notebook:3772fffc4aa4
      volumes:
        - /home/sproell/Jupyter_Workdir:/home/jovyan/work
        - ./certificates:/etc/ssl/notebook
      ports:
        - 10000:8888
      container_name:   datascience-notebook-container
      environment:
        USE_HTTPS: "yes"
        GEN_CERT: 'yes'
      command: "start-notebook.sh"
```
When you try to set a password, the error  appears

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/jovyan/.jupyter/jupyter_notebook_config.json'
```
The reason is clearly that the config file is not created in the image. When the service tries to write a new password to the config file, it fails in error. My work around is pretty simple too. I go inside the container with `docker exec -it datascience-notebook-container bash` and create a new config with `jupyter notebook --generate-config`. While I am there, I also set a new password with `jupyter notebook password`. That solves the issue.

**Comment by parente** [comment]
I believe this was closed by #671 . Thanks again @stefanproell
